Index: http.py
===================================================================
--- http.py	(revision 57)
+++ http.py	(working copy)
@@ -4,6 +4,7 @@
 
 import cStringIO
 import dpkt
+import gzip
 
 def parse_headers(f):
     """Return dict of HTTP headers parsed from a file object."""
@@ -55,6 +56,10 @@
     else:
         # XXX - need to handle HTTP/0.9
         body = ''
+
+    if headers.get('content-encoding', '').lower() == 'gzip':
+        body = gzip.Gzip(body).decompress()
+
     return body
 
 class Message(dpkt.Packet):
